File: src/js/core/src/Prng.ts
import { Fnv1a } from './Prng/Fnv1a.js';
import { Mulberry32 } from './Prng/Mulberry32.js';
import type { Range } from './StyleDefinition.js';

/**
 * Key-based pseudorandom number generator.
 *
 * Each method takes a key that, combined with the seed, produces a
 * deterministic value. The same seed + key always yields the same result,
 * regardless of call order.
 */
export class Prng {
  #seed: string;

  constructor(seed: string) {
    this.#seed = seed;
  }

  /**
   * Picks a single item from `items` deterministically. Returns `undefined`
   * for an empty list. Duplicate values (by string representation) are
   * collapsed before picking so that input order and duplication do not
   * affect the result.
   */
  pick<T>(key: string, items: readonly T[]): T | undefined {
    if (items.length === 0) {
      return undefined;
    }

    if (items.length === 1) {
      return items[0];
    }

    const unique = this.#uniqueByCodePoint(items);

    if (unique.length === 1) {
      return unique[0];
    }

    const sorted = unique.sort(this.#compareByCodePoint);
    const index = Math.floor(this.getValue(key) * sorted.length);

    return sorted[index];
  }

  /**
   * Picks an item from `entries` proportional to its weight. Duplicate items
   * (by string representation) are collapsed before picking — only the
   * first occurrence's weight is kept. When all weights are zero, falls
   * back to an unweighted {@link pick}. Returns `undefined` for an empty
   * list.
   */
  weightedPick<T>(
    key: string,
    entries: readonly (readonly [T, number])[],
  ): T | undefined {
    if (entries.length === 0) {
      return undefined;
    }

    if (entries.length === 1) {
      return entries[0][0];
    }

    const unique = this.#uniqueByCodePoint(entries, (e) => String(e[0]));
    const sorted = unique.sort((a, b) => this.#compareByCodePoint(a[0], b[0]));
    const totalWeight = sorted.reduce((sum, e) => sum + e[1], 0);

    if (totalWeight === 0) {
      return this.pick(
        key,
        sorted.map((e) => e[0]),
      );
    }

    const threshold = this.getValue(key) * totalWeight;

    let cumulative = 0;

    for (const [item, weight] of sorted) {
      cumulative += weight;

      if (threshold < cumulative) {
        return item;
      }
    }

    return sorted[sorted.length - 1][0];
  }

  /**
   * Returns `true` with the given probability (0–100, default 50).
   */
  bool(key: string, likelihood: number = 50): boolean {
    return this.getValue(key) * 100 < likelihood;
  }

  /**
   * Returns a deterministic float in `range`, rounded to four decimal places.
   * With `range.step > 0`, the result is drawn uniformly from
   * `{ min + i*step | 0 ≤ i ≤ floor((max - min) / step) }`, so both endpoints
   * of an evenly-divisible range are equally likely. Non-positive or absent
   * step means continuous. `min`/`max` are sorted internally, so a reversed
   * pair is tolerated.
   */
  float(key: string, range: Range): number {
    const min = Math.min(range.min, range.max);
    const max = Math.max(range.min, range.max);
    const step = range.step ?? 0;
    let value: number;

    if (step > 0) {
      const buckets = Math.floor((max - min) / step) + 1;
      const i = Math.floor(this.getValue(key) * buckets);
      value = min + i * step;
    } else {
      value = min + this.getValue(key) * (max - min);
    }

    return Math.round(value * 10000) / 10000;
  }

  /**
   * Returns a deterministic integer in `range`. `min`/`max` are sorted
   * internally, so a reversed pair is tolerated. `range.step` is accepted
   * for symmetry with {@link float} but ignored — integers already step by 1.
   */
  integer(key: string, range: Range): number {
    const min = Math.min(range.min, range.max);
    const max = Math.max(range.min, range.max);

    return Math.floor(this.getValue(key) * (max - min + 1)) + min;
  }

  /**
   * Fisher-Yates shuffle with chained Mulberry32 state. Duplicate values
   * (by string representation) are collapsed before shuffling, so a
   * caller's slice off the front cannot accidentally produce a repeated
   * value.
   */
  shuffle<T>(key: string, items: readonly T[]): T[] {
    if (items.length <= 1) {
      return [...items];
    }

    const result = this.#uniqueByCodePoint(items).sort(
      this.#compareByCodePoint,
    );
    const prng = new Mulberry32(Fnv1a.hash(this.#seed + ':' + key));

    for (let i = result.length - 1; i > 0; i--) {
      const j = Math.floor(prng.nextFloat() * (i + 1));
      const temp = result[i];

      result[i] = result[j];
      result[j] = temp;
    }

    return result;
  }

  /**
   * Returns a single float in `[0, 1)` derived from `seed:key`. The same
   * seed/key pair always produces the same value.
   */
  getValue(key: string): number {
    return new Mulberry32(Fnv1a.hash(this.#seed + ':' + key)).nextFloat();
  }

  /**
   * Deduplicates by string representation, keeping the first occurrence.
   * Mirrors the cross-language sort key used by {@link #compareByCodePoint}
   * so that JS and PHP collapse the same set of inputs. `keyFn` lets
   * callers (e.g. {@link weightedPick}) extract the sort key from a
   * compound element.
   */
  #uniqueByCodePoint<T>(
    items: readonly T[],
    keyFn: (item: T) => string = String,
  ): T[] {
    const seen = new Set<string>();
    const result: T[] = [];

    for (const item of items) {
      const repr = keyFn(item);

      if (!seen.has(repr)) {
        seen.add(repr);
        result.push(item);
      }
    }

    return result;
  }

  /**
   * Cross-language deterministic sort: compare by UTF-16 code units of
   * the string representation. Every language can reproduce this by
   * converting values to strings and comparing code unit by code unit.
   */
  #compareByCodePoint<T>(a: T, b: T): number {
    const sa = String(a);
    const sb = String(b);

    if (sa < sb) {
      return -1;
    }

    if (sa > sb) {
      return 1;
    }

    return 0;
  }
}

File: src/dart/core/lib/src/prng/prng.dart
/// Key-based pseudorandom number generator and its primitives (FNV-1a,
/// Mulberry32). Each method takes a key that, combined with the seed,
/// produces a deterministic value: the same seed + key always yields the
/// same result, regardless of call order.
library;

import 'dart:math' as math;

import '../range.dart';
import '../utils/number.dart';
import 'fnv1a.dart';
import 'mulberry32.dart';

/// Key-based pseudorandom number generator.
///
/// Each method takes a key that, combined with the seed, produces a
/// deterministic value. The same seed + key always yields the same result,
/// regardless of call order.
class Prng {
  final String _seed;

  Prng(String seed) : _seed = seed;

  /// Returns a single float in `[0, 1)` derived from `seed:key`. The same
  /// seed/key pair always produces the same value.
  double getValue(String key) {
    return Mulberry32(fnv1aHash('$_seed:$key')).nextFloat();
  }

  /// Picks a single item from [items] deterministically. Returns `null` for
  /// an empty list. Duplicate values (by string representation) are collapsed
  /// before picking so that input order and duplication do not affect the
  /// result.
  T? pick<T extends Object>(String key, List<T> items) {
    if (items.isEmpty) {
      return null;
    }

    if (items.length == 1) {
      return items[0];
    }

    final unique = _uniqueByCodePoint(items);

    if (unique.length == 1) {
      return unique[0];
    }

    // After dedupe all sort keys are distinct, so the instability of
    // Dart's `List.sort` cannot be observed.
    unique.sort(_compareByCodePoint);
    final index = (getValue(key) * unique.length).floor();

    return unique[index];
  }

  /// Picks a key from [weights] proportional to its weight. When all weights
  /// are zero, falls back to an unweighted [pick]. Returns `null` for an
  /// empty map.
  String? weightedPick(String key, Map<String, double> weights) {
    if (weights.isEmpty) {
      return null;
    }

    if (weights.length == 1) {
      return weights.keys.first;
    }

    // `String.compareTo` is the UTF-16 code-unit order the JS reference
    // sorts by; the map's insertion order must not leak in.
    final sorted = weights.keys.toList()..sort();

    // Sum in sorted-key order to match the JS reduce-over-sorted parity:
    // float addition is non-associative, so insertion order would diverge.
    var totalWeight = 0.0;

    for (final k in sorted) {
      totalWeight += weights[k]!;
    }

    if (totalWeight == 0) {
      return pick(key, sorted);
    }

    final threshold = getValue(key) * totalWeight;
    var cumulative = 0.0;

    for (final k in sorted) {
      cumulative += weights[k]!;

      if (threshold < cumulative) {
        return k;
      }
    }

    // Safety fallback for float accumulation edges.
    return sorted.last;
  }

  /// Returns `true` with the given probability (0–100, default 50).
  ///
  /// Named `boolean` because `bool` is a reserved type name in Dart; this is
  /// the JS port's `bool`.
  bool boolean(String key, [double likelihood = 50]) {
    return getValue(key) * 100 < likelihood;
  }

  /// Returns a deterministic float in [range], rounded to four decimal
  /// places. With `range.step > 0`, the result is drawn uniformly from
  /// `{ min + i*step | 0 ≤ i ≤ floor((max - min) / step) }`, so both
  /// endpoints of an evenly-divisible range are equally likely. Non-positive
  /// or absent step means continuous. `min`/`max` are sorted internally, so
  /// a reversed pair is tolerated.
  double float(String key, Range range) {
    final min = math.min(range.min, range.max);
    final max = math.max(range.min, range.max);
    final step = range.step ?? 0;
    double value;

    if (step > 0) {
      final buckets = ((max - min) / step).floorToDouble() + 1;
      final i = (getValue(key) * buckets).floorToDouble();
      value = min + i * step;
    } else {
      value = min + getValue(key) * (max - min);
    }

    // JS Math.round rounds half toward +∞; Dart's `round()` would diverge
    // for negative halves. Multiply, round, then divide — per-operation,
    // matching the other ports.
    return roundHalfUp(value * 10000) / 10000;
  }

  /// Returns a deterministic integer in [range]. `min`/`max` are sorted
  /// internally, so a reversed pair is tolerated. `range.step` is accepted
  /// for symmetry with [float] but ignored — integers already step by 1.
  int integer(String key, Range range) {
    final min = math.min(range.min, range.max);
    final max = math.max(range.min, range.max);

    // Compute in doubles like the JS reference and convert at the end; the
    // floored value plus min is always integral and well below 2^53.
    return ((getValue(key) * (max - min + 1)).floorToDouble() + min).truncate();
  }

  /// Fisher-Yates shuffle with chained Mulberry32 state. Duplicate values
  /// (by string representation) are collapsed before shuffling, so a
  /// caller's slice off the front cannot accidentally produce a repeated
  /// value. Always returns a new list.
  List<T> shuffle<T extends Object>(String key, List<T> items) {
    if (items.length <= 1) {
      return [...items];
    }

    final result = _uniqueByCodePoint(items)..sort(_compareByCodePoint);

    // One chained generator with the same `seed:key` derivation as
    // [getValue], so the first draw equals getValue(key); the remaining
    // n-2 draws continue the same state.
    final prng = Mulberry32(fnv1aHash('$_seed:$key'));

    for (var i = result.length - 1; i > 0; i--) {
      final j = (prng.nextFloat() * (i + 1)).floor();
      final temp = result[i];

      result[i] = result[j];
      result[j] = temp;
    }

    return result;
  }
}

/// JS `String(item)` semantics for dedupe/sort keys: identity for strings,
/// the JS-compatible number formatter for nums (`fontWeight` lists are
/// numeric). Never raw `double.toString()`, which Dart renders as `400.0`
/// where JS prints `400`.
String _jsString(Object item) {
  return item is num ? jsNumString(item) : item.toString();
}

/// Deduplicates by string representation, keeping the first occurrence.
/// Mirrors the cross-language sort key used by [_compareByCodePoint] so that
/// every port collapses the same set of inputs.
List<T> _uniqueByCodePoint<T extends Object>(List<T> items) {
  final seen = <String>{};
  final result = <T>[];

  for (final item in items) {
    if (seen.add(_jsString(item))) {
      result.add(item);
    }
  }

  return result;
}

/// Cross-language deterministic sort: compare by UTF-16 code units of the
/// string representation (JS string `<` compares code units). Dart strings
/// are UTF-16, so [String.compareTo] is exactly this — it agrees with
/// code-point order for all BMP text and differs only for
/// supplementary-plane characters, e.g. "😀" (0xD83D 0xDE00) sorts before
/// U+E000 because 0xD83D < 0xE000.
int _compareByCodePoint(Object a, Object b) {
  return _jsString(a).compareTo(_jsString(b));
}
